### PR TITLE
[red-knot] Fallback to attributes on types.ModuleType if a symbol can't be found in locals or globals

### PR DESCRIPTION
**Pull request by AlexWaygood**
## Summary

All modules in Python are instances of `types.ModuleType`. [Various attributes on `types.ModuleType` in typeshed](https://github.com/python/typeshed/blob/783171b9f7412823ed1a531e9c84feee755f8303/stdlib/types.pyi#L331-L344) represent "implicit globals" that are part of the namespace of every module in Python, even if not explicitly defined. As such, before declaring that a symbol is unbound (and in fact, before looking the symbol up in the `builtins` scope) we should look the symbol up as an attribute on `types.ModuleType`. The only attributes on `types.ModuleType` that are _not_ present as implicit globals are `__dict__`, `__getattr__` and `__init__`.

## Test Plan

Markdown test added as `crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md`. We also get to remove a false-positive error from the `tomllib` benchmark 🎉

### CODE REVIEW

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 1]
```diff
@@ -2359,12 +2361,24 @@ impl<'db> TypeInferenceBuilder<'db> {
...
-            } else {
-                global_symbol_ty(self.db, self.file, name)
+            if !file_scope_id.is_global() {
+                ty = global_symbol_ty(self.db, self.file, name);
             };
 
+            // Still possibly unbound? All modules are instances of `types.ModuleType`;
+            // look it up there (with a few very special exceptions)
+            if ty.may_be_unbound(self.db)
+                && !matches!(&**name, "__dict__" | "__init__" | "__getattr__")
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/infer.rs:2371]: Hmmm... since we anyway having to hardcode a denylist of attributes _not_ to lookup on `types.ModuleType`, I suppose I could instead have an allowlist of attributes we _should_ lookup on `types.ModuleType`.

That has the downside that new implicit globals might be added to the `types.ModuleType` stub in the future and we wouldn't automatically recognise them as such unless we updated our hardcoded allowlist. But it might be much better for performance.

Comment (MichaReiser) [reply crates/red_knot_python_semantic/src/types/infer.rs:2371]: Interestingly, the regression is for the incremental benchmark. The cold benchmark is barely affected. Not sure why that is. Some local profiling could help to understand the difference. 

To me the main question here is: How common is that we reach unbound here? If it's not very common in valid code (unless we expect specific attributes), than the performance penalty in actual code should be low.

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types/infer.rs:2371]: One issue is that we have to lookup symbols on `ModuleType` before we look them up in the builtins scope. But it's realistically much more likely that a symbol is going to resolve to a symbol in `builtins` than that it's going to be an attribute on `ModuleType` :/ So I think it's quite common that we'd reach `Unbound` here and then have to continue to the next part of the algorithm where we look it up in `builtins.pyi`.

Comment (MichaReiser) [reply crates/red_knot_python_semantic/src/types/infer.rs:2371]: True. But let's only do that if this is the root cause

Comment (carljm) [reply crates/red_knot_python_semantic/src/types/infer.rs:2371]: Sure -- but I would strongly suspect that the root cause has _something_ to do with looking up `ModuleType` in typeshed and looking up a member on it, since that's really the only change in this diff. Currently this diff does that work for every lookup of a builtin (which are common), we could instead do it only for those names where it is relevant.

So my next steps here would be to a) implement a hardcoded version of the name allowlist (because it's quick and easy) and see if that fixes the regression (I strongly suspect it will), and if so, then turn it into a Salsa query allowlist instead, so we remain robust to typeshed changes.

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 2]
```diff
@@ -2329,6 +2329,8 @@ impl<'db> TypeInferenceBuilder<'db> {
             .expect("Symbol table should create a symbol for every Name node")
             .is_bound();
 
+        let mut ty = Type::Unbound;
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:2332]: I think this change to the structure of the function is not necessary in this PR, and I have mixed feelings about it. One the one hand I see that it removes a couple `else` clauses and arguably makes the function structure easier to understand. On the other hand, I think it might move this function further away from the structure it will need to have when we remove `Type::Unbound`, at which point there won't be any reasonable "default" value to set `ty` to?

**Review by carljm — CHANGES_REQUESTED** [review]
Looks good, but I do think we need to resolve the regression here, if at all possible; this is too much of an edge-case to pay that much perf for.

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 2 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types/infer.rs:2332]: Happy to revert this change, it was just a driveby thing to improve readability

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 1 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types/infer.rs]: Confirmed that hardcoding an allowlist rather than a denylist does fix the perf regression!

File: crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+# error: [unresolved-reference]
+# revealed: Unbound
+reveal_type(__init__)
+```
+
+## Accessed as attributes
+
+`ModuleType` attributes can also be accessed as attributes
+on module-literal inhabitants:
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md:54]: Minor nit: I would say module literal types here. From the perspective of the type checker, type inhabitants are theoretical runtime objects, which we don't have access to, since we aren't the runtime. So we can't access an attribute on an inhabitant, only on a type.

**Review by carljm — COMMENTED** [review]
Looks great! I assume you are still planning to try replacing the hard coded allowlist with a salsa query based on actual attributes found in typeshed, just noticed one minor terminology nit.

File: crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md (hunk 3 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md:54]: yes, fair enough

**Comment by AlexWaygood** [comment]
Okay, I've switched from the hardcoded allowlist to a Salsa-cached generated allowlist! Locally I'm consistently seeing a regression of 1-2% on the incremental benchmark, but [Codspeed reports the benchmarks are neutral on this PR](https://codspeed.io/astral-sh/ruff/branches/moduletype-attrs), so maybe it's just noise.

In order to obtain a list of symbols that typeshed declares in the body of `types.ModuleType`, I had to add a new `SymbolsFlag` member, `IS_DECLARED`. We had ways of querying whether symbols in a `SymbolTable` had been _defined_ in that scope, but no way of querying whether symbols in a `SymbolTable` had been _declared_ in that scope -- and for a stub file, the latter is what's important.

I've also fixed several edge cases regarding accessing members as attributes on imported modules. In some cases, whether a module "defines" a member can vary depending on whether you access the member as a global variable from inside the module or as an attribute on the module object after it's been imported elsewhere.

TL;DR: ready for re-review!

File: crates/red_knot_python_semantic/src/semantic_index/symbol.rs [hunk 4]
```diff
@@ -47,17 +47,22 @@ impl Symbol {
...
+    }
 }
 
 bitflags! {
     #[derive(Copy, Clone, Debug, Eq, PartialEq)]
     struct SymbolFlags: u8 {
         const IS_USED         = 1 << 0;
-        const IS_BOUND      = 1 << 1;
+        const IS_BOUND        = 1 << 1;
+        const IS_DECLARED     = 1 << 2;
```
Comment (MichaReiser) [inline crates/red_knot_python_semantic/src/semantic_index/symbol.rs:61]: Could you add some comment explaining `IS_DECLARED`'s meaning?

File: crates/red_knot_python_semantic/src/types.rs [hunk 5]
```diff
@@ -88,9 +88,64 @@ fn symbol_ty<'db>(db: &'db dyn Db, scope: ScopeId<'db>, name: &str) -> Type<'db>
         .unwrap_or(Type::Unbound)
 }
 
-/// Shorthand for `symbol_ty` that looks up a module-global symbol by name in a file.
+/// Return a list of the symbols that typeshed declares in the scope of `types.ModuleType`
+#[salsa::tracked(return_ref)]
+fn module_type_symbols<'db>(db: &'db dyn Db) -> Box<[Box<str>]> {
```
Comment (MichaReiser) [inline crates/red_knot_python_semantic/src/types.rs:93]: How many entries do we expect `module_type_symbols` to contain? Returning a list is probably fine performance wise if it is a very small number of elements. Otherwise it might be worth considering a `HashSet` (has the downside that it always requires hashing the entire input string) or a `BTreeSet`

File: crates/red_knot_python_semantic/src/types.rs [hunk 6]
```diff
@@ -88,9 +88,64 @@ fn symbol_ty<'db>(db: &'db dyn Db, scope: ScopeId<'db>, name: &str) -> Type<'db>
...
+    let Some(module_type) = KnownClass::ModuleType
+        .to_class(db)
+        .into_class_literal_type()
+    else {
+        return Box::default();
+    };
+
+    let module_type_scope = module_type.body_scope(db);
+    let module_type_symbol_table = symbol_table(db, module_type_scope);
```
Comment (MichaReiser) [inline crates/red_knot_python_semantic/src/types.rs:102]: I don't understand enough about `ModuleType` to assess this and it might be a very dumb question so I'm just gonna raise it here and leave it up to you.

The symbol table only does lexicographical scoping, meaning it doesn't contain resolved entries from star imports. Is that a problem here or could it become a problem if typeshed adds such a star import?

File: crates/red_knot_python_semantic/src/types.rs [hunk 7]
```diff
@@ -88,9 +88,64 @@ fn symbol_ty<'db>(db: &'db dyn Db, scope: ScopeId<'db>, name: &str) -> Type<'db>
...
+
+    // `__dict__` and `__init__` are very special members that can be accessed as attributes
+    // on the module when imported, but cannot be accessed as globals *inside* the module.
+    //
+    // `__getattr__` is even more special: it doesn't exist at runtime, but typeshed includes it
+    // to reduce false positives associated with functions that dynamically import modules
+    // and return `Instance(types.ModuleType)`. We should ignore it for any known module-literal type.
+    let module_type_symbols: Box<[Box<str>]> = module_type_symbol_table
```
Comment (MichaReiser) [inline crates/red_knot_python_semantic/src/types.rs:110]: It could be more efficient to preserve `Name` over `Box<str>` to avoid allocating for very small strings (note, this is probably not worth it when using a `Set` because `contains` then requires creating a new Name)

File: crates/red_knot_python_semantic/src/types.rs [hunk 8]
```diff
@@ -88,9 +88,64 @@ fn symbol_ty<'db>(db: &'db dyn Db, scope: ScopeId<'db>, name: &str) -> Type<'db>
...
+        .filter(|symbol| symbol.is_declared())
+        .map(|symbol| &**symbol.name())
+        .filter(|symbol_name| !matches!(*symbol_name, "__dict__" | "__getattr__" | "__init__"))
+        .map(Box::from)
+        .collect();
+
+    // Make sure it contains declared symbols in the `types.ModuleType` scope,
+    // but not other random symbols that are only *referenced* in the `types.ModuleType` scope
+    debug_assert!(module_type_symbols.contains(&Box::from("__name__")));
+    debug_assert!(!module_type_symbols.contains(&Box::from("property")));
```
Comment (MichaReiser) [inline crates/red_knot_python_semantic/src/types.rs:121]: These assertions feel more like a test than a debug assertion to me.

**Review by MichaReiser — COMMENTED** [review]
It seems unfortunate to me that we now have to track `DECLARED` just for this one use case.

File: crates/red_knot_python_semantic/src/types.rs (hunk 5 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:93]: The query just returns all symbols declared in the body of this class definition in typeshed (excluding 3 very special attributes): https://github.com/python/typeshed/blob/c225ac758781c0599889b320c0a4f334366c5907/stdlib/types.pyi#L331. Currently that's a list of length 6; it might change in the future (which is why it's a salsa query rather than a hardcoded list) but I do not expect it to ever be very long. For this reason I did not go for a set, as I assumed that the cost of hashing the strings would be more costly than it was worth.

File: crates/red_knot_python_semantic/src/types.rs (hunk 6 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:102]: This query just returns a list of the symbols declared in the body of [this class in typeshed](https://github.com/python/typeshed/blob/c225ac758781c0599889b320c0a4f334366c5907/stdlib/types.pyi#L331); I don't think star imports are relevant here :-)

File: crates/red_knot_python_semantic/src/types.rs (hunk 8 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:121]: That's fair. I couldn't see a way of doing an mdtest asserting this, but I suppose I can still just write an old-fashioned unittest for the function!

File: crates/red_knot_python_semantic/src/types.rs (hunk 7 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:110]: Oh good call — I forgot that `Name` avoided allocating for small strings by wrapping a `CompactString`!

File: crates/red_knot_python_semantic/src/semantic_index/symbol.rs (hunk 4 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/semantic_index/symbol.rs:66]: I added a link to Carl's essay at the top of https://github.com/astral-sh/ruff/blob/main/crates/red_knot_python_semantic/src/semantic_index/use_def.rs

**Comment by AlexWaygood** [comment]
> It seems unfortunate to me that we now have to track `DECLARED` for all symbols just for this one use case.

I would wager that this information will come in useful in other contexts in the future (though I can't right now say _what_ -- it's just an intuition I have).

_However_, one alternative way of tackling this would be to write a Python script that inspects the AST of the typeshed stub for `types.ModuleType` and generates a Rust file that contains a function that returns `true` if a given symbol is declared in the body of `types.ModuleType`. (This would be similar to the approach we take in https://github.com/astral-sh/ruff/blob/main/scripts/generate_known_standard_library.py.) We could run the script as part of every typeshed sync, to make sure that the generated function is always up to date with the typeshed stubs.

This would be a more complicated than the current solution. But it would have the following advantages:
- We could probably use the [`typeshed_client`](https://github.com/JelleZijlstra/typeshed_client) library in the Python script, which means that the script itself probably wouldn't be _that_ complicated
- The generated Rust function would likely be more performant than the current solution that uses a Salsa query, because we wouldn't have to look anything up in the cache: the specific symbols that are declared on `types.ModuleType` would just be hardcoded into the generated Rust function.
- We wouldn't have to track `DECLARED` for all symbols just for this one use case

I'm not sure whether this would be better or worse on net than the solution I currently have in this PR -- but it's an alternative to consider, anyway!

**Comment by MichaReiser** [comment]
> I would wager that this information will come in useful in other contexts in the future (though I can't right now say what -- it's just an intuition I have).

Yeah. I'm not a fan of it but I don't consider it blocking. I guess an alternative to this would be to look at the class type and iterate over its members. I don't think we support it today, but I suspect that's a capability we want long-term.

File: crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+reveal_type(__package__)  # revealed: @Todo
+reveal_type(__spec__)  # revealed: @Todo
+
+# TODO: generics
+reveal_type(__path__)  # revealed: @Todo
+
+# TODO: this should probably be added to typeshed; not sure why it isn't?
+# error: [unresolved-reference]
+# revealed: Unbound
+reveal_type(__doc__)
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md:25]: Probably because it's on `builtins.object`?

The awkward thing is that I think `__doc__` might be the only attribute on `object` that is accessible in every module as a global name.

I think all of them except for `__module__` are accessible on module objects as an attribute.

You'll know better than I whether it's likely we can get `__doc__` added specifically to `ModuleType` (despite it already being on `object`) to help support fixing this TODO with fewer special cases, or whether we should instead just bite the bullet and handle it as a special case.

File: crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+# error: [unresolved-reference]
+# revealed: Unbound
+reveal_type(__dict__)
+
+# error: [unresolved-reference]
+# revealed: Unbound
+reveal_type(__init__)
+```
+
+## Accessed as attributes
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md:51]: `__class__` is another one defined on `object` that is accessible on `ModuleType` instances, but not as a module global name.

`__module__` is an odd special case in that typeshed says it exists on `object` but it does not actually exist on module objects. Probably not worth modeling this LSP violation, it's fine if we just let it exist on all objects?

File: crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+## `types.ModuleType.__dict__` takes precedence over global variable `__dict__`
+
+It's impossible to override the `__dict__` attribute of `types.ModuleType`
+instances from inside the module; we should prioritise the attribute in
+the `types.ModuleType` stub over a variable named `__dict__` in the module's
+global namespace:
+
+```py path=foo.py
+__dict__ = "foo"
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md:90]: can we also test here that accessing `__dict__` within the module, subsequent to this, does get `Literal["foo"]`? But `from foo import __dict__ as d` gives you the module dictionary?

**Review by carljm — APPROVED** [review]
Looks great! A few comments that I think are worth at least testing and/or mentioning in TODOs, even if we don't make code changes for them right now.

> an alternative to this would be to look at the class type and iterate over its members. I don't think we support it today, but I suspect that's a capability we want long-term.

I agree that we'll want this, and I think the best / most efficient way to implement this would be to iterate over symbols in the symbol table that are defined/declared in the class, using `IS_DEFINED` and `IS_DECLARED` flags. So I would flip it around and say this is not an alternative approach, rather it's another likely future use case for this flag.

File: crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md (hunk 9 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md:25]: Yeah... I think it would be fine to add `__doc__` to `ModuleType` in typeshed if we add a comment saying why we have the "redundant" override from `object`... I'll try to put up a typeshed PR today...

File: crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md (hunk 10 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md:51]: > `__module__` is an odd special case in that typeshed says it exists on `object` but it does not actually exist on module objects. Probably not worth modeling this LSP violation, it's fine if we just let it exist on all objects?

*Sigh*. `__module__` [really](https://github.com/python/typeshed/pull/8789) [shouldn't](https://github.com/python/typeshed/pull/8787) be listed as an instance attribute on `object`; it's just obviously incorrect. But I'm not going to restart that conversation 🙃

I think this is not worth special-casing in red-knot; we should just accept it as an inaccuracy in type inference that we have to live with unless and until it's fixed in typeshed.

File: crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md (hunk 11 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md:90]: > But `from foo import __dict__ as d` gives you the module dictionary?

Woah, TIL. That's nuts.

File: crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md (hunk 9 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md:23]: Filed https://github.com/python/typeshed/pull/12918

File: crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md (hunk 11 above)
Comment (carljm) [reply crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md:92]: From imports are just syntax sugar for an attribute access!
